posts/comparisons-with-other-projects.md

```diff
@@ -0,0 +1,69 @@
+however this project does in a way hint that the magma project is a good idea! it is also generic over different architectures and uses automatic verification condition generators
+
+## ivy http://microsoft.github.io/ivy/language.html
+only first order
+
+## bedrock
+Proprietary! It's essential systems like this aren't only controlled by corporations and governments.
+http://adam.chlipala.net/papers/BedrockICFP13/BedrockICFP13.pdf
+https://plv.csail.mit.edu/bedrock/
+https://bedrocksystems.com/products/
+The original purely research version of bedrock is yet another project that is promising for the magma project, since it shows that verified *macros* are possible and tractable. However it's still stuck in coq and therefore slow and obtuse.
+
+need to look at xcap paper and other references in the bedrock paper
+
+
+## https://github.com/mit-plv/bedrock2
+
+at first glance it would seem this project is directly trying to do the same thing, but it seems to have dramatically different goals
+
+- no self-hosting
+- not creating a full proof checker
+- building a language at a relatively high level of abstraction rather than bare metal assembly
+
+however that team will definitely have a ton of useful insights to provide! the work is similar enough there's got to be enough overlap to make working together productive.
+
+
+## mm0
+
+b-system
+
+
+
+Existing research around formal methods and program verification, such as in the deepspec project, I believe focuses extremely foolishly on old software workflows and tools. C and LLVM, despite being extremely powerful tools that profoundly advanced computing, are still necessarily "legacy" tools, and so only very clunkily fit into formal verification. Even Rust, as modern as it is, wasn't ever designed with formal verification in mind from the beginning, and inherited many possibly unhelpful assumptions, such as the operating system syscall model, LLVM itself, and the C++ memory model.
+
+The ability to fully verify any program down to the metal completely unlocks the kinds of systems and abstractions we can build! The very act of building such a powerful tool removes all non-negotiable barriers in our way.
+
+Formally verifying the correctness of legacy systems after the fact is necessarily much more difficult than developing new tools from first principles with verification in mind. The only reason I can think of for the post-hoc philosophy is one of terrified pragmatism, where researchers and engineers are too scared to rethink layers that "seem to be working". This seems foolish to me, since we don't *actually* have any confidence those old layers are actually correct. If you start from the bottom and produce fully verified foundations, and every layer you stack on top is itself verified, I conjecture you can move much faster than trying to avoid or work-around bugs in existing legacy foundations.
```

Comment:
@blainehansen This is true if you can ever get from the bottom to the top in one (huge) step. My point about using a subset of an existing language as an intermediate step was simply to bootstrap the project. Since you'd presumably ultimately want or need to write magma in magma, you have to bootstrap somehow. Once you've learned enough about the problem space and "proven" the approach, you are free to throw away the bootstrapping and re-bootstrap formally from scratch.

Anyway, I'll be happy to be proven wrong, but I suspect it won't be in my lifetime (I'm 61).

Reply:
@glyn no fair you're looking at my incomplete scribblings! haha I think I actually wrote those words months ago and just commented them out yesterday. 

But I've been thinking a lot about what you've been saying, I just responded to your issue about the matter :)